### ISSUE TITLE
.NET : Couldn't parse Assembly Analyzer results with Mono-complete installed

### ISSUE THREAD

**Issue by Arnab035** [dependency-check/DependencyCheck#1715]
### Reporting Bugs/Errors

I am running the dependency check program to scan a Windows Image (`microsoft/nanoserver:latest`). The dependency check program is running inside a Linux Container. The Linux Container runs on a Windows Host (i.e. the Linux container is a LCOW container variant). The Windows host is a Windows server 2019 edition and I am using the latest Docker-EE version too (Docker 18.09.1).

I have installed Mono-complete inside the container to facilitate the dependency-check program to work with .NET binaries. 

```
root@*********:/tmp# mono --version
Mono JIT compiler version 4.2.1 (Debian 4.2.1.102+dfsg2-7ubuntu4)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        LLVM:          supported, not enabled.
        GC:            sgen
```

I run the dependency-check program inside the container with the following flags specified.

```
/usr/local/bin/deepaudit -ti http://<ip-address>:<port> -mount-path /fenced/mnt/host -scan-type dotnet -i microsoft/nanoserver -proxy http://<ip-address>:<proxy-port>
```
The program runs for a while after which it exits "fatally" and I get errors in the log which says that the "Assembly analyzer could not be parsed."

```
DEBUG - Created '20' parsers.
2019-02-02 13:46:21,020 org.owasp.dependencycheck.reporting.VelocityLoggerRedirect:72
DEBUG - Velocimacro : "velocimacro.library" is not set.  Trying default library: VM_global_library.vm
2019-02-02 13:46:21,022 org.owasp.dependencycheck.reporting.VelocityLoggerRedirect:72
DEBUG - Velocimacro : Default library not found.
2019-02-02 13:46:21,022 org.owasp.dependencycheck.reporting.VelocityLoggerRedirect:72
DEBUG - Velocimacro : allowInline = true : VMs can be defined inline in templates
2019-02-02 13:46:21,022 org.owasp.dependencycheck.reporting.VelocityLoggerRedirect:72
DEBUG - Velocimacro : allowInlineToOverride = false : VMs defined inline may NOT replace previous VM definitions
2019-02-02 13:46:21,022 org.owasp.dependencycheck.reporting.VelocityLoggerRedirect:72
DEBUG - Velocimacro : allowInlineLocal = false : VMs defined inline will be global in scope if allowed.
2019-02-02 13:46:21,022 org.owasp.dependencycheck.reporting.VelocityLoggerRedirect:72
DEBUG - Velocimacro : autoload off : VM system will not automatically reload global library macros
2019-02-02 13:46:22,314 org.owasp.dependencycheck.App:204
ERROR - Couldn't parse Assembly Analyzer results (GrokAssembly)
2019-02-02 13:46:22,315 org.owasp.dependencycheck.utils.Settings:624
DEBUG - Deleting ALL temporary files from `/tmp/dctemp77139b1b-9219-4ded-b4ff-f257e25ea9f6`
2019-02-02 13:46:22,340 org.owasp.dependencycheck.App:70
DEBUG - Exit code: -14
```

Since I already have `mono-complete` installed, I did not expect this error to arise. 

What could be the reason for this error ? Are there any workarounds to this problem ?

**Comment by tylercamp** [comment]
We've also reproduced this issue in Docker containers running on a tomcat image (based on `debian:stretch-slim` image). We install `mono-complete`, and a few other things based on the official mono docker image ([here](https://github.com/mono/docker/blob/d68573b0640b3e191bf41b64745a6f0683a9c17a/5.18.0.225/slim/Dockerfile) and [here](https://github.com/mono/docker/blob/d68573b0640b3e191bf41b64745a6f0683a9c17a/5.18.0.225/Dockerfile)).

We're running Dependency Check 1.4.3 as its own JAR for scanning inputs and get the logs below. This occurs for 12/19 .NET test projects we use, such as [mojoPortal](https://github.com/i7MEDIA/mojoportal), [NEsper](https://github.com/espertechinc/nesper), [NUnit](https://github.com/nunit/nunit), [Serenity](https://github.com/volkanceylan/Serenity), and the [WingTip Toys](https://code.msdn.microsoft.com/Getting-Started-with-221c01f5?cdn_id=2013-12-16-001) demo.

```
[INFO] Central analyzer disabled
[INFO] Analysis Started
[INFO] Finished Archive Analyzer (8 seconds)
[INFO] Finished File Name Analyzer (0 seconds)
[INFO] Finished Nuspec Analyzer (0 seconds)
[Fatal Error] :1:1: Content is not allowed in prolog.
[ERROR] ----------------------------------------------------
[ERROR] Failed to read the Assembly Analyzer results. On some systems mono-runtime and mono-devel need to be installed.
[ERROR] ----------------------------------------------------
[WARN] An error occurred while analyzing '/tmp/dctemp852cbfde-1797-4611-9df7-f2cdda2e3b59/check1864555632684741858tmp/1/Serene1-SQL Server/Serene1/Serene1.Script/bin/Debug/Saltarelle.jQuery.dll' (Assembly Analyzer).
[Fatal Error] :1:1: Content is not allowed in prolog.
[ERROR] ----------------------------------------------------
[ERROR] Failed to read the Assembly Analyzer results. On some systems mono-runtime and mono-devel need to be installed.
[ERROR] ----------------------------------------------------
[WARN] An error occurred while analyzing '/tmp/dctemp852cbfde-1797-4611-9df7-f2cdda2e3b59/check1864555632684741858tmp/1/Serene1-SQL Server/Serene1/Serene1.Script/bin/Debug/Saltarelle.jQuery.UI.dll' (Assembly Analyzer).
[Fatal Error] :1:1: Content is not allowed in prolog.
[ERROR] ----------------------------------------------------
[ERROR] Failed to read the Assembly Analyzer results. On some systems mono-runtime and mono-devel need to be installed.
[ERROR] ----------------------------------------------------
[WARN] An error occurred while analyzing '/tmp/dctemp852cbfde-1797-4611-9df7-f2cdda2e3b59/check1864555632684741858tmp/1/Serene1-SQL Server/Serene1/Serene1.Script/bin/Debug/Saltarelle.Linq.dll' (Assembly Analyzer).
[Fatal Error] :1:1: Content is not allowed in prolog.
[ERROR] ----------------------------------------------------
[ERROR] Failed to read the Assembly Analyzer results. On some systems mono-runtime and mono-devel need to be installed.
[ERROR] ----------------------------------------------------
[WARN] An error occurred while analyzing '/tmp/dctemp852cbfde-1797-4611-9df7-f2cdda2e3b59/check1864555632684741858tmp/1/Serene1-SQL Server/Serene1/Serene1.Script/bin/Debug/Saltarelle.Web.dll' (Assembly Analyzer).
[Fatal Error] :1:1: Content is not allowed in prolog.
[ERROR] ----------------------------------------------------
[ERROR] Failed to read the Assembly Analyzer results. On some systems mono-runtime and mono-devel need to be installed.
[ERROR] ----------------------------------------------------
[WARN] An error occurred while analyzing '/tmp/dctemp852cbfde-1797-4611-9df7-f2cdda2e3b59/check1864555632684741858tmp/1/Serene1-SQL Server/packages/Serenity.CodeGenerator.1.4.6/tools/sergen.exe' (Assembly Analyzer).
[INFO] Finished Assembly Analyzer (3 seconds)
[INFO] Finished Dependency Merging Analyzer (0 seconds)
[INFO] Finished Version Filter Analyzer (0 seconds)
[INFO] Finished Hint Analyzer (0 seconds)
[INFO] Created CPE Index (4 seconds)
[INFO] Skipping CPE Analysis for npm
[INFO] Finished CPE Analyzer (5 seconds)
[INFO] Finished False Positive Analyzer (0 seconds)
[INFO] Finished NVD CVE Analyzer (1 seconds)
[INFO] Finished Vulnerability Suppression Analyzer (0 seconds)
[INFO] Finished Dependency Bundling Analyzer (0 seconds)
[INFO] Analysis Complete (19 seconds)
[ERROR] Couldn't parse Assembly Analyzer results (GrokAssembly)
[ERROR] Couldn't parse Assembly Analyzer results (GrokAssembly)
[ERROR] Couldn't parse Assembly Analyzer results (GrokAssembly)
[ERROR] Couldn't parse Assembly Analyzer results (GrokAssembly)
[ERROR] Couldn't parse Assembly Analyzer results (GrokAssembly)
```

**Comment by tylercamp** [comment]
Running `GrokAssembly` manually got us:

```
Can't find custom attr constructor image: /opt/codedx/analysis-files/10/14/test/PHOENIXGP.TrainingPortal.dll mtoken: 0x0a000014 due to: Could not load file or assembly 'WebActivator, Version=1.5.3.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
<?xml version="1.0" encoding="utf-8"?><assembly><company>UNKNOWN</company><product>PHOENIXGP.TrainingPortal</product><version>1.0.0.0</version><fullname>PHOENIXGP.TrainingPortal, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null</fullname></assembly>
```

Seems the initial error message is also being sent to the XML parser followed by the actual XML. The mentioned assembly `WebActivator` is in the same location as the assembly being scanned.

**Comment by tylercamp** [comment]
Was able to mostly fix the issue by using Mono v4.6.2.7 specifically (the official LTS version)

@Arnab035 Try using that mono version instead

Seems like there are two issues

1. DependencyCheck reading that error message in the XML output (not sure if GrokAssembly is properly separating stdout/stderr)
2. GrokAssembly requiring that all dependencies be resolvable (even though it's just getting the full name of the assembly) which isn't necessary

#1 would help but wouldn't fix this issue, #2 may require some other reflection library to avoid that unnecessary work

**Comment by jeremylong** [comment]
5.0.0-M2 which will be released (hopefully) soon will require dotnet core to be installed.

**Comment by jeremylong** [comment]
We missed closing this ticket when 5.0.0 was released.  5.0-5.3 uses dotnet 2.x. With the soon to be released 5.4.0 we will move to dotnet 3.1.
